### PR TITLE
gh-94808: Cover `%p` in `PyUnicode_FromFormat`

### PR DESCRIPTION
**Pull request by sobolevn**
Source: https://github.com/python/cpython/blob/4f523a70543a79a3bbca9bf2ce9abb8fad0aaad2/Objects/unicodeobject.c#L2531-L2552

Internet says that some systems can return representation with `0x`, some with `0X`, and some with none. It is reflected in the implementation, but cannot be reflected in tests.

* Issue: gh-94808

### CODE REVIEW

**Comment by sobolevn** [comment]
And we got a failure: MacOS passes locally and on CI, but other platforms do not:

```
 ======================================================================
FAIL: test_from_format (test.test_unicode.CAPITest.test_from_format)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\test_unicode.py", line 2817, in test_from_format
    self.assertEqual(len(p_format1), 11)
AssertionError: 10 != 11
```

**Review by kumaraditya303 — COMMENTED** [review]
I think a regex test would be better here.

**Comment by sobolevn** [comment]
Good idea, thank you.

File: Lib/test/test_unicode.py [hunk 1]
```diff
@@ -2807,6 +2807,19 @@ def check_format(expected, format, *args):
...
+        # test %p
+        # We cannot test the exact result,
+        # because it returns a hex representation of a C pointer,
+        # which is going to be different each time. But, we can test the format.
+        p_format_regex = r'0x[a-zA-Z0-9]{8,}'
+        p_format1 = PyUnicode_FromFormat(b'%p', 'abc')
+        self.assertIsInstance(p_format1, str)
+        self.assertRegex(p_format1, p_format_regex)
+
+        p_format2 = PyUnicode_FromFormat(b'repr=%p', '123456', None, b'xyz')
```
Comment (JelleZijlstra) [inline Lib/test/test_unicode.py]: What does this case test? I don't see what it adds over the previous case.

Comment (sobolevn) [reply Lib/test/test_unicode.py]: It has:
- Multiple arguments (while `p_format1` only has one arg)
- All arguments have different types
- New types are tested: `bytes` and `None`

Comment (JelleZijlstra) [reply Lib/test/test_unicode.py]: Aren't the additional arguments ignored though, because there is only one `%p` in the format string?

Comment (sobolevn) [reply Lib/test/test_unicode.py]: I agree, tests can be better :)
Can you please take a look at the updated version?

It now has both:
- Extra types
- Ignored arguments

**Review by JelleZijlstra — APPROVED** [review]

File: Lib/test/test_unicode.py (hunk 1 above)
Comment (JelleZijlstra) [reply Lib/test/test_unicode.py]: Thanks!

**Comment by sobolevn** [comment]
From https://github.com/python/cpython/actions/runs/3200613350/jobs/5227726848

```
AssertionError: Regex didn't match: '0x[a-zA-Z0-9]{8,} 0x[a-zA-Z0-9]{1,} 0x[a-zA-Z0-9]{8,}' not found in '0x60400191c290 0x(nil) 0x60400191c230'
```

This is quite interesting. Is `0x(nil)` valid?

**Comment by JelleZijlstra** [comment]
Interesting! I think that comes from libc since I can't find code in CPython that would do this.

**Comment by JelleZijlstra** [comment]
I think ctypes translates None into NULL, which is reasonable.

File: Lib/test/test_unicode.py [hunk 2]
```diff
@@ -2807,6 +2807,25 @@ def check_format(expected, format, *args):
         check_format('repr=abc',
                      b'repr=%V', 'abc', b'xyz')
 
+        # test %p
+        # We cannot test the exact result,
+        # because it returns a hex representation of a C pointer,
+        # which is going to be different each time. But, we can test the format.
+        p_format_regex = r'0x[a-zA-Z0-9]{8,}'
```
Comment (JelleZijlstra) [inline Lib/test/test_unicode.py]:
```suggestion
        p_format_regex = r'^0x[a-zA-Z0-9]{8,}$'
```
Make sure it matches the whole string

**Comment by JelleZijlstra** [comment]
I guess the pointers are shorter on s390. I'll adjust the regex.

**Pull request merged**